nixos/modules/system/boot/systemd/initrd.nix

```diff
@@ -332,7 +332,10 @@ in {
   config = mkIf (config.boot.initrd.enable && cfg.enable) {
     system.build = { inherit initialRamdisk; };
 
-    boot.initrd.availableKernelModules = [ "autofs4" ]; # systemd needs this for some features
+    boot.initrd.availableKernelModules = [
+      "autofs4"           # systemd needs this for some features
+      "tpm-tis" "tpm-crb" # systemd-cryptenroll
```

Comment:
Can we maybe make this optional? Or conditional on if the kernel has this option set?

I have some custom kernel configs that really don't need this module and also use the new initrd and now the initrd build fails with `modprobe: FATAL: Module tpm-tis not found in directory […]`.

Reply:
I guess we can have a new option under `boot.initrd.luks.devices.<name>` that indicates whether a LUKS makes use of cryptenroll-based TPM2/FIDO2 unlocking, and only insert the modules and extra files when a volume with those options enabled exists. However, this can be confusing for non-systemd-stage1 users, especially when `boot.initrd.luks.devices.<name>.fido2` already exists.

Reply:
I'd prefer if disabling these kernel modules would an opt-in config option.

We can't determine at evaluation time if there's a fido luks slot configured.

Reply:
fair enough. iirc we _can_ determine at eval time if the kernel even has these modules enabled. maybe using `system.requiredKernelConfig`?

Reply:
looking at the kernel config, these options/modules don't even seem to be available on non-x86 systems or systems without ACPI, so I'm really not sure they should be enabled by default

Reply:
Hmmh, that'd be a problem. Can we check if the kernel has these modules configured, and only include them in that case?

Reply:
I think something like this might work:
```diff
--- a/nixos/modules/system/boot/systemd/initrd.nix                                                                                                                                    
+++ b/nixos/modules/system/boot/systemd/initrd.nix
@@ -334,8 +334,11 @@ in {
               
     boot.initrd.availableKernelModules = [
       "autofs4"           # systemd needs this for some features
-      "tpm-tis" "tpm-crb" # systemd-cryptenroll              
-    ];                                                                                    
+    ]                                                                                     
+      # systemd-cryptenroll                                                                                                                                                          
+      ++ lib.optional (config.boot.kernelPackages.kernel.config.isEnabled "TCG_TIS") "tpm-tis"
+      ++ lib.optional (config.boot.kernelPackages.kernel.config.isEnabled "TCG_CRB") "tpm-crb"                                                                     
+    ;                                                                                     
                                
     boot.initrd.systemd = {                                                               
       initrdBin = [pkgs.bash pkgs.coreutils cfg.package.kmod cfg.package] ++ config.system.fsPackages;                                                                              
```
but only if we actually set these options and don't rely on kernel defaults. Maybe.

For some reason `nixosTests.systemd-initrd-luks-fido2` and `nixosTests.systemd-initrd-luks-tpm2` build even with:
```
     boot.initrd.availableKernelModules = [
       "autofs4"           # systemd needs this for some features
     ];
```
so I'm not sure how to test.

Reply:
The VM test passes without adding any extra modules but on (some?) actual hardware it doesn't.